src/net/peer/connect/connect.rs

```diff
@@ -81,170 +83,161 @@ quick_error! {
-                        .map_err(ConnectError::Io)
-                }?;
-                direct_connections
-                    .select(hole_punch_connections)
-                    .into_boxed()
-            } else {
-                direct_connections.into_boxed()
-            };
+    let direct_incoming =
+        handshake_incoming_connections(our_connect_request, peer_rx, their_id, name_hash);
+    let handle_copy = handle.clone();
+    all_connections
+        .select(direct_incoming)
+        .first_ok()
+        .map_err(ConnectError::AllConnectionsFailed)
+        .and_then(move |(socket, their_uid)| {
+            peer::from_handshaken_socket(&handle_copy, socket, their_uid, CrustUser::Node)
+                .map_err(ConnectError::Io)
+        })
+        .into_boxed()
+}
 
-            let handle0 = handle.clone();
-            connections
-                .map_err(SingleConnectionError::Io)
-                .map(move |(stream, addr)| {
-                    Socket::wrap_tcp(&handle0, stream, addr)
-                })
-                .and_then(move |socket| {
-                    socket
-                        .send((0, HandshakeMessage::Connect(our_connect_request.clone())))
-                        .map_err(SingleConnectionError::Socket)
-                })
-                .map(move |socket| {
-                    socket
-                        .into_future()
-                        .map_err(|(err, _socket)| SingleConnectionError::Socket(err))
-                        .with_timeout(Duration::from_secs(TIMEOUT_SEC), &handle)
-                        .and_then(|res| res.ok_or(SingleConnectionError::TimedOut))
-                })
-                .buffer_unordered(128)
-                .and_then(move |(msg_opt, socket)| match msg_opt {
-                    None => Err(SingleConnectionError::ConnectionDropped),
-                    Some(HandshakeMessage::Connect(connect_request)) => {
-                        validate_connect_request(their_uid, name_hash, &connect_request)?;
-                        Ok((socket, their_uid))
-                    }
-                    Some(_msg) => Err(SingleConnectionError::UnexpectedMessage),
-                })
-        };
+fn connect_directly(
+    evloop_handle: &Handle,
+    addrs: Vec<SocketAddr>,
+) -> BoxStream<TcpStream, SingleConnectionError> {
+    stream::futures_unordered(
+        addrs
+            .into_iter()
+            .map(|addr| TcpStream::connect(&addr, evloop_handle))
+            .collect::<Vec<_>>(),
+    ).map_err(SingleConnectionError::Io)
+        .into_boxed()
+}
 
-        let all_connections = direct_incoming.select(other_connections);
+fn handshake_incoming_connections<UID: Uid>(
+    our_connect_request: ConnectRequest<UID>,
+    conn_rx: UnboundedReceiver<ConnectMessage<UID>>,
+    their_id: UID,
+    name_hash: NameHash,
+) -> BoxStream<(Socket<HandshakeMessage<UID>>, UID), SingleConnectionError> {
+    conn_rx
+        .map_err(|()| unreachable!())
```

Comment:
I'd prefer to use the `Unbounded{Sender,Reciever}` from `future-utils` just so we don't have to have `unreachable!`s in our code. Although it's a shame not use a more standard type.

Reply:
Makes sense, I'd agree with that. Except this requires more changes in different parts of code (`demux`) since `conn_rx` is passed to us by `demux`.
Anyway, I believe this change is out of scope of this PR.